File: app_pojav_zh/src/main/java/com/movtery/background/BackgroundManager.java
package com.movtery.background;

import net.kdt.pojavlaunch.PojavZHTools;
import net.kdt.pojavlaunch.Tools;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.InputStream;
import java.util.Map;
import java.util.Properties;

public class BackgroundManager {
    public static File FILE_BACKGROUND_PROPERTIES = new File(Tools.DIR_GAME_HOME, "background.properties");

    public static Properties getProperties() {
        if (!FILE_BACKGROUND_PROPERTIES.exists()) {
            return getDefaultProperties();
        }
        Properties properties = new Properties();
        try (
                InputStream is = new FileInputStream(FILE_BACKGROUND_PROPERTIES)
        ) {
            properties.load(is);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        return properties;
    }

    private static Properties getDefaultProperties() {
        Properties properties = new Properties();
        properties.setProperty(BackgroundType.MAIN_MENU.name(), "null");
        properties.setProperty(BackgroundType.SETTINGS.name(), "null");
        properties.setProperty(BackgroundType.CUSTOM_CONTROLS.name(), "null");
        properties.setProperty(BackgroundType.IN_GAME.name(), "null");

        saveProperties(properties);
        return properties;
    }
    
    private static void saveProperties(Properties properties) {
        if (!PojavZHTools.DIR_BACKGROUND.exists()) PojavZHTools.DIR_BACKGROUND.mkdirs();

        try {
            properties.store(new FileWriter(FILE_BACKGROUND_PROPERTIES), "Pojav ZH Background Properties File");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static void saveProperties(Map<BackgroundType, String> map) {
        Properties properties = new Properties();
        properties.setProperty(BackgroundType.MAIN_MENU.name(),
                (map.get(BackgroundType.MAIN_MENU) == null ? "null" : map.get(BackgroundType.MAIN_MENU)));
        properties.setProperty(BackgroundType.SETTINGS.name(),
                (map.get(BackgroundType.SETTINGS) == null ? "null" : map.get(BackgroundType.SETTINGS)));
        properties.setProperty(BackgroundType.CUSTOM_CONTROLS.name(),
                (map.get(BackgroundType.CUSTOM_CONTROLS) == null ? "null" : map.get(BackgroundType.CUSTOM_CONTROLS)));
        properties.setProperty(BackgroundType.IN_GAME.name(),
                (map.get(BackgroundType.IN_GAME) == null ? "null" : map.get(BackgroundType.IN_GAME)));

        saveProperties(properties);
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/feature/background/BackgroundManager.kt
package com.movtery.pojavzh.feature.background

import android.annotation.SuppressLint
import android.content.Context
import android.widget.ImageView
import androidx.core.content.ContextCompat
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.DrawableImageViewTarget
import com.movtery.pojavzh.utils.PathAndUrlManager
import com.movtery.pojavzh.utils.file.FileTools.Companion.mkdirs
import com.movtery.pojavzh.utils.image.ImageUtils.Companion.isImage
import net.kdt.pojavlaunch.R
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.util.Properties

class BackgroundManager private constructor(val context: Context) {
    companion object {
        private val FILE_BACKGROUND_PROPERTIES: File = File(PathAndUrlManager.DIR_GAME_HOME, "background.properties")
        @SuppressLint("StaticFieldLeak")
        private var sBackgroundManager: BackgroundManager? = null

        @JvmStatic
        fun setContext(context: Context) {
            sBackgroundManager = BackgroundManager(context)
        }

        @JvmStatic
        fun getInstance() = sBackgroundManager
    }

    private val defaultProperties: Properties
        get() {
            val properties = Properties()
            properties.setProperty(BackgroundType.MAIN_MENU.name, "null")
            properties.setProperty(BackgroundType.CUSTOM_CONTROLS.name, "null")
            properties.setProperty(BackgroundType.IN_GAME.name, "null")

            saveProperties(properties)
            return properties
        }

    val properties: Properties
        get() {
            if (!FILE_BACKGROUND_PROPERTIES.exists()) {
                return defaultProperties
            }
            val properties = Properties()
            try {
                FileReader(FILE_BACKGROUND_PROPERTIES).use { fileReader ->
                    properties.load(fileReader)
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

            return properties
        }

    fun setBackgroundImage(
        backgroundType: BackgroundType,
        backgroundView: ImageView
    ) {
        backgroundView.setImageDrawable(
            ContextCompat.getDrawable(context, R.color.background_app)
        )

        val backgroundImage = getBackgroundImage(backgroundType) ?: return

        Glide.with(context).load(backgroundImage)
            .override(backgroundView.width, backgroundView.height)
            .centerCrop()
            .into(DrawableImageViewTarget(backgroundView))
    }

    private fun getBackgroundImage(backgroundType: BackgroundType): File? {
        val pngName = properties[backgroundType.name] as String?
        if (pngName == null || pngName == "null") return null

        val backgroundImage = File(PathAndUrlManager.DIR_BACKGROUND, pngName)
        if (!backgroundImage.exists() || !isImage(backgroundImage)) return null
        return backgroundImage
    }

    private fun saveProperties(properties: Properties) {
        val dirBackground = PathAndUrlManager.DIR_BACKGROUND
        if (!dirBackground!!.exists()) mkdirs(dirBackground)

        try {
            properties.store(
                FileWriter(FILE_BACKGROUND_PROPERTIES),
                "Pojav Zenith Horizon Background Properties File"
            )
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun saveProperties(map: Map<BackgroundType?, String?>) {
        val properties = Properties()
        properties.setProperty(
            BackgroundType.MAIN_MENU.name,
            map[BackgroundType.MAIN_MENU] ?: "null"
        )
        properties.setProperty(
            BackgroundType.CUSTOM_CONTROLS.name,
            map[BackgroundType.CUSTOM_CONTROLS] ?: "null"
        )
        properties.setProperty(
            BackgroundType.IN_GAME.name,
            map[BackgroundType.IN_GAME] ?: "null"
        )

        saveProperties(properties)
    }
}